### DISCUSSION TITLE
Feature Request: Static Build Compilation

### DISCUSSION THREAD

**Discussion by Gosokan** [stickz/rtorrent#15] [General]
Can you add an automation script to build static rtorrent, similiar to [this repo](https://github.com/chros73/rtorrent-ps-ch/blob/master/docs/DebianInstallFromSourceTheEasyWay.rst#compiling-vanilla-build)? So the compiled results can be used/copied to any machine (as long as the operating system is the same version).

**Comment by stickz** [comment]
Hello @Gosokan, I don't have time to do this anytime soon. My focus right now is improving the rTorrent client further and releasing ruTorrent version 5. But if you have any questions about how to construct it, I can point you in the right direction. And review the script.

**A few notes for a static build:**
- LTO is optional. It will improve performance slightly. Give it a try, but don't stress out if it doesn't work. Skip it and move on.
- It's required to build with GCC. GCC 13.3 is recommended at this moment in time. But don't stress if you can't easily obtain it.
- The GNU c++ library (libstdc++) is required. Do not use the clang libc++. We're still resolving issues with libc++.
- You can use ether glibc or musl for your c library. Both are supported by this project. glibc has more thorough testing and should be your first option. Then musl your second option.
- You must include curl within your static build. It's highly recommended to use version 8 for higher levels of software performance.
- You must build curl with c-ares. It's very important for TCP trackers. Version 1.31.0 is highly recommended at this moment in time. The built in DNS caching feature greatly improves performance. https://github.com/c-ares/c-ares/releases/tag/v1.31.0
- Use our versions of UDNS and xmlrpc-c on the readme page. Build the software with level 3 optimizations and all the make and configure options we recommend. It will greatly improve the performance and stability of the software solution.

**Reply by Gosokan** [reply]
I just tried compiling a static build, but it failed when compiling rtorrent 
The version of GCC on Ubuntu that I use is 11.4.0
libstdc++6 is already the newest version (12.3.0-1ubuntu1~22.04)

```
:~/sources/rtorrent-stickz/rtorrent-0.9.8# make CXXFLAGS="-O3"
...
/usr/bin/ld: rpc/libsub_rpc.a(xmlrpc.o): in function `rpc::xmlrpc_to_target(_xmlrpc_env*, _xmlrpc_value*)':
xmlrpc.cc:(.text+0x1ef9): undefined reference to `torrent::hash_string_from_hex_c_str(char const*, torrent::HashString&)'
/usr/bin/ld: rpc/libsub_rpc.a(xmlrpc.o): in function `rpc::xmlrpc_call_command(_xmlrpc_env*, _xmlrpc_value*, void*)':
xmlrpc.cc:(.text+0x385b): undefined reference to `torrent::Object::swap(torrent::Object&)'
/usr/bin/ld: xmlrpc.cc:(.text+0x39da): undefined reference to `torrent::Object::swap(torrent::Object&)'
/usr/bin/ld: xmlrpc.cc:(.text+0x3af7): undefined reference to `torrent::Object::swap(torrent::Object&)'
/usr/bin/ld: xmlrpc.cc:(.text+0x3b1a): undefined reference to `torrent::Object::swap(torrent::Object&)'
/usr/bin/ld: utils/libsub_utils.a(directory.o): in function `utils::Directory::update(int) [clone .cold]':
directory.cc:(.text.unlikely+0x10d): undefined reference to `torrent::input_error::initialize(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'

/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:564: rtorrent] Error 1
make[3]: Leaving directory '/root/sources/rtorrent-stickz/rtorrent-0.9.8/src'
make[2]: *** [Makefile:635: all-recursive] Error 1
make[2]: Leaving directory '/root/sources/rtorrent-stickz/rtorrent-0.9.8/src'
make[1]: *** [Makefile:456: all-recursive] Error 1
make[1]: Leaving directory '/root/sources/rtorrent-stickz/rtorrent-0.9.8'
make: *** [Makefile:388: all] Error 2
```

**Reply by stickz** [reply]
Did you use the correct prefix targets when running configure for everything? The ones on the readme (`./configure --prefix=/usr`) are for a shared build. You'll need to create your own prefix to reference the dependencies.

I can see here there are undefined references to the libtorrent dependency. Did you build the one from our GitHub? Are you accidently referencing the system version (or anther version) of libtorrent instead of our version?

Take a look at this line from one of our GitHub actions runners. The LDFLAGS are overwritten to link directly to the correct version.
https://github.com/stickz/rtorrent/blob/master/.github/scripts/build_whatbox.sh#L23

**Reply by Gosokan** [reply]
Thank you for the instructions about `LDFLAGS`, I have successfully compiled rtorrent (not fully static binary but it can be moved/copied to another computer without needing to recompile).

**Reply by Auska** [reply]
I am preparing to build fully static binaries using the x86_64-linux-musl toolchain.I have already installed ncurses 6.5, but I am still encountering errors.
```
../../rak/priority_queue_default.h:56:7: note: in C++11 destructors default to 'noexcept'
./../display/attributes.h: At global scope:
./../display/attributes.h:31:32: error: 'A_NORMAL' was not declared in this scope; did you mean 'FP_NORMAL'?
   31 |   static const int a_normal  = A_NORMAL;
      |                                ^~~~~~~~
      |                                FP_NORMAL
./../display/attributes.h:32:32: error: 'A_BOLD' was not declared in this scope
   32 |   static const int a_bold    = A_BOLD;
      |                                ^~~~~~
./../display/attributes.h:33:32: error: 'A_REVERSE' was not declared in this scope
   33 |   static const int a_reverse = A_REVERSE;
      |                                ^~~~~~~~~
download_list.cc: In member function 'void core::DownloadList::clear()':
```
ncurses 6.5:
```
./configure --prefix=$PREFIX --enable-colorfgbg --enable-overwrite --enable-pc-files --without-ada --host=x86_64-unknown-linux-musl
```

**Reply by Auska** [reply]
I've finally succeeded, [those who need it can download it directly.](https://github.com/Auska/rtorrent-static)

**Reply by roman-lautenschlager** [reply]
> I've finally succeeded, [those who need it can download it directly.](https://github.com/Auska/rtorrent-static)

Hi. Could you please share the steps / script to build it? Thanks a lot.

**Reply by silasmariusz** [reply]
<img width="585" height="1266" alt="IMG_3022" src="https://github.com/user-attachments/assets/214f4365-8556-4675-8d39-cff6fbc60574" />
<img width="1289" height="516" alt="image" src="https://github.com/user-attachments/assets/465229a1-fa20-4526-9411-cf8db138a2e6" />
Is it worth for server except embedded system like rpi or low performance PC?
